screenwriter/screenwriter.go

```diff
@@ -0,0 +1,70 @@
+package screenwriter
+
+import (
+	"io"
+	"strings"
+)
+
+// ScreenWriter allows writing messages to the screen and then overwriting those messages.
+type ScreenWriter struct {
+	writer          io.StringWriter
+	previousMessage string
+}
+
+// New creates a new ScreenWriter which is ready to be used.
+func New(writer io.StringWriter) *ScreenWriter {
+	return &ScreenWriter{
+		writer: writer,
+	}
+}
+
+func (w *ScreenWriter) clearPrevious() error {
+	if w.previousMessage == "" {
+		return nil
+	}
+
+	lines := strings.Split(w.previousMessage, "\n")
+
+	// Move the cursor to the start of the line
+	finalLine := lines[len(lines)-1]
+	finalLineLength := len(finalLine)
+	err := moveCursor(w.writer, directionLeft, finalLineLength)
+	if err != nil {
+		return err
+	}
+
+	// We need to clear the last line in case the string does not end with a "\n"
+	err = clearLine(w.writer)
+	if err != nil {
+		return err
+	}
```

Comment:
Still not clear why we need this (why does it matter if the last line ends with `\n`?), but not a blocker.

Reply:
If the string ends with `\n` then we don't need to clear the last line because the last line is empty/blank. If the string doesn't end with `\n` then that means there are characters which need to be cleared from the display (unless they're ANSI control codes, or non-printing characters).

Reply:
Would this be simpler if we just cleared the empty lines, too?